### DISCUSSION TITLE
Using pdb in OOAnalyzer

### DISCUSSION THREAD

**Discussion by dimitrijejankov** [cmu-sei/pharos#171] [Q&A]
Dear OOAnalyzer team,

Suppose I had access to more information about the types, like a .pdb file. Is it possible to explicitly bake in these 
constraints into the prolog rules? If not could you give me a few hits on how to extend it? :)

**Comment by sei-eschwartz** [comment]
PDBs likely contain all the information you need.  We actually use PDBs to extract ground truth for testing OOAnalyzer.  Is there a reason you can't simply import the PDB into Ghidra/IDA?

**Reply by dimitrijejankov** [reply]
the pdb does not match the binary but I know the information inside is correct. Therefore I only have the types and method present but I can not relate them an address.

**Reply by sei-eschwartz** [reply]
I can't think of any uses for this off the top of my head.  OOAnalyzer does almost all of its reasoning based off addresses (methods, vftables, etc.) so if you can't relate the information to an address it won't be much help.

I wonder if you could use the information in the PDB to fuzzy match methods between the PDB and your executable.  For example, you can extrapolate the addresses and sizes of methods from the PDB.  You could use OOAnalyzer (or some other tool) to do the same from your executable.  And then try to align the methods based on order and size.  If the binary and PDB are similar, then many of the method sizes will be the same and presumably in the same order.

**Reply by sei-ccohen** [reply]
I think another important question is whether you also have the other executable (the one that _does_ match the PDB).  If you do, I would recommend the following procedure, where the "old" executable matches the PDB, and the "new" executable does not.

1. Load the old executable and PDB in Ghidra to get some symbol "ground truth".
2. Generate function hashes using either CERT Kaiju or pharos fn2hash, for the old executable.
3. Generate function hashes for the new executable using either CERT Kaiju or pharos fn2hash,
4. Match functions by their hash, starting with exact hashes, then PIC hashes, then some of the other hashes.

You'll need to watch the uniqueness of the hashes because there will some cases of multiple functions using the same hashes (especially for very small functions).   Those are best sorted out with a little control flow graph analysis.  The Pharos dumpmasm tool might help (it exports a control flow graph) if you want to script it.  Otherwise, I'd just plow through in Ghidra checking your results manually as you go.

Good luck!

**Reply by dimitrijejankov** [reply]
I don't have that one. Thanks for the replies :)

**Reply by sei-ccohen** [reply]
Github seems to be flaking out a little bit for me because your replies don't always show.    I agree that it would be nice to be able to supplement facts from another source, whether it be a PDB or manual analysis from a related executable.  We've talked about this a little, but it gets very complicated because you don't really know what's the same and what has changed between the files, and because OOAnalyzer work on addresses, it's hard to tie the user-supplied facts to the discovered classes.

OOAnalyzer uses the RTTI, and should do pretty well on the classes that have RTTI.  It'll be the classes without RTTI that are likely to give you the most trouble.   It's a hard problem, which is why we've been building a tool to help. ;-)

**Reply by dimitrijejankov** [reply]
I kicked of OOAnalyzer on the binary, and the prolog part crashed. The log said to send that there might be a problem with the rules. Here is my working directory if you guys want to debug this. 
https://drive.google.com/file/d/17_oDVPboGUULuW4hajL4ywnJehgazx9L/view?usp=sharing

**Reply by edmcman** [reply]
I don't have permission to view the file on drive.

Can you create a new issue for this problem?

**Reply by dimitrijejankov** [reply]
just did that :)
